HW4.py

```diff
@@ -0,0 +1,316 @@
+                p_value = "{:.6e}".format(
+                    linear_regression_model_fitted.pvalues[1]
+                )  # noqa
+                # Continuous Predictor by Categorical Response
+                # Plot the figure
+                fig = px.histogram(x=column, y=y)
+                fig.update_layout(
+                    title=f"Variable: {feature_name}: (t-value={t_value}) (p-value={p_value})",  # noqa
+                    xaxis_title=f"Variable: {feature_name}",
+                    yaxis_title="y",
+                )
+                fig.show()
+        # create a list of each variable type
+        predictor_type.append(v_type)
+
+    print("***Difference with mean table***")
+
+    # Create difference with mean table
+
+    # create a temp table df_bin to store raw data
+    df = pd.DataFrame(data.data, columns=data.feature_names)
+    df["target"] = data["target"]
+    count_row = df.shape[0]
+    count_column = df.shape[1]
+
+    predictor_list = []
+    for sublist in X:
+        for item in sublist:
+            predictor_list.append(item)
+    p_min = min(predictor_list)
+    p_max = max(predictor_list)
+    p_range = p_max - p_min
+    n_of_bin = 10
+    bin_width = p_range / n_of_bin
+
+    bin_list = [p_min]
+    s = 0
+    while s < p_max:
+        s += bin_width
+        bin_list.append(s)
+
+    df_bin = pd.DataFrame({"predictor": pd.Series(predictor_list)})
+    df_bin["LowerBin_UpperBin"] = pd.cut(
+        x=pd.Series(predictor_list), bins=bin_list, include_lowest=True
+    )
+    bincenter = []
+    for bin_n in df_bin["LowerBin_UpperBin"]:
+        bincenter.append(bin_n.mid)
+    df_bincenters = pd.DataFrame({"BinCenters": pd.Series(bincenter)})
+    df_bin["BinCenters"] = df_bincenters
+    df_bin["response"] = pd.Series(np.repeat(y, count_column - 1))
+
+    # Groupby df_bin table to create a Difference with mean table
+
+    df_bin_groupby = df_bin.groupby(("LowerBin_UpperBin"), as_index=False).agg(
+        bin_mean=pd.NamedAgg(column="predictor", aggfunc="mean"),
+        bin_count=pd.NamedAgg(column="predictor", aggfunc="count"),
+    )
+
+    bin_center_list = []
+    for bin_center in df_bin_groupby["LowerBin_UpperBin"]:
+        bin_center_list.append(bin_center.mid)
+
+    df_bin_groupby["BinCenter"] = pd.Series(bin_center_list)
+
+    PopulationMean = (np.sum(X) + np.sum(y)) / (count_row * count_column)
+    df_bin["PopulationMean"] = PopulationMean
+    df_bin_groupby["PopulationMean"] = PopulationMean
+
+    MeanSquaredDiff = (
+        df_bin_groupby["bin_mean"] - df_bin_groupby["PopulationMean"]
+    ) ** 2
+    df_bin_groupby["MeanSquaredDiff"] = MeanSquaredDiff
+
+    print(df_bin_groupby)
+
+    # Square the difference, sum them up and divide by number of bins
+    print(
+        f"THE unWeighted NUMBER IS : {df_bin_groupby['MeanSquaredDiff'].sum() / n_of_bin}"  # noqa
+    )
+
+    fig = px.bar(df_bin_groupby, x="BinCenter", y="bin_count")

```

Comment:
Check julian's example in the slides. This graph needs the part you have, but also a line plot and a population mean line overlaid.

Reply:
updated